core/utils/trie.h

```diff
@@ -34,26 +40,47 @@ class Trie {
     }
 
     bool leaf;
+    bool prefix;
+    T val;
     std::unique_ptr<Node> children[256];
   };
 
   Trie() : root_() {}
-  Trie(const Trie& t) : root_(t.root_) { }
+  Trie(const Trie& t) : root_(t.root_) {}
 
-  // Inserts a string into the trie
-  void Insert(const std::string& key);
+  // Inserts a string into the trie, associating the key
+  // with the value.
+  void Insert(const std::string& key, const T val);
```

Comment:
You should probably have `val` be passed as a const reference since T is likely not a reference type.  (Here and for other methods.)